bitcoin/test/fuzz/miniscript_random.cpp

```diff
@@ -0,0 +1,417 @@
+
+/**
+ * Generate a Miniscript node based on the fuzzer's input.
+ * Note this does not attempt to produce a well-typed (let alone safe) Miniscript node.
+ */
+NodeRef GenNode(FuzzedDataProvider& provider, const miniscript::Type typ, const size_t recursion_depth);
+
+/**
+ * Generate a pseudorandom node. If invalid, return NULL.
+ * Used to cut-through as a node with invalid subs will never be valid.
+ */
+NodeRef GenValidNode(FuzzedDataProvider& provider, const miniscript::Type typ, const size_t recursion_depth)
+{
+    const NodeRef node = GenNode(provider, typ, recursion_depth);
+    if (!node || !node->IsValid()) return {}; // TODO!
+    return node;
+}
+
+//! Generate a vector of miniscript nodes of the given types.
+std::vector<NodeRef> MultiNode(FuzzedDataProvider& provider, const std::vector<miniscript::Type>& types,
+                               const size_t recursion_depth)
+{
+    std::vector<NodeRef> subs;
+    for (size_t i = 0; i < types.size(); ++i) {
+        const NodeRef sub = GenValidNode(provider, types[i], recursion_depth);
+        if (!sub) return {};
+        subs.push_back(sub);
+    }
+    return subs;
+}
+
+//! Generate a node with pseudorandom subs of the given types.
+NodeRef MultiSubs(FuzzedDataProvider& provider, const NodeType node_type, const std::vector<miniscript::Type>& subtypes,
+                  const size_t recursion_depth)
+{
+    const auto subs = MultiNode(provider, subtypes, recursion_depth);
+    if (subs.empty()) return {};
+    return MakeNodeRef(node_type, subs);
+}
+
+NodeRef GenNode(FuzzedDataProvider& provider, const miniscript::Type typ, const size_t recursion_depth) {
+    if (recursion_depth >= MAX_NESTED_DEPTH) return {};
+
+    if (typ << "B"_mst) {
+        switch (provider.ConsumeIntegralInRange<size_t>(0, 19)) {
+            case 0: return MakeNodeRef(provider.ConsumeBool() ? NodeType::JUST_0 : NodeType::JUST_1);
+            case 1: {
+                const uint32_t k{provider.ConsumeIntegralInRange<uint32_t>(1, CTxIn::SEQUENCE_LOCKTIME_DISABLE_FLAG - 1)};
+                return MakeNodeRef(provider.ConsumeBool() ? NodeType::OLDER : NodeType::AFTER, k);
+            }
+            case 2: {
+                const size_t hashtype = provider.ConsumeIntegralInRange<size_t>(0, 3);
+                const size_t index = provider.ConsumeIntegralInRange<size_t>(0, 255);
+                switch (hashtype) {
+                    case 0: return MakeNodeRef(NodeType::SHA256, TEST_DATA.sha256[index]);
+                    case 1: return MakeNodeRef(NodeType::RIPEMD160, TEST_DATA.ripemd160[index]);
+                    case 2: return MakeNodeRef(NodeType::HASH256, TEST_DATA.hash256[index]);
+                    case 3: return MakeNodeRef(NodeType::HASH160, TEST_DATA.hash160[index]);
+                }
+            }
+            case 3: {
+                if (const NodeRef sub = GenValidNode(provider, "K"_mst, recursion_depth + 1)) {
```

Comment:
If you don't make this `const`, you can use `Vector(std::move(sub))` below, avoiding a shared_ptr copy (here, and elsewhere).